GHA: Automated transfer of unsolved project notebooks (#5)
datatrail-jhu/DataTrail_Projects: 03_Cleaning_the_Data/countries_project.Rmd

## Set up directories

> Here we are going to make a data directory if it doesn't already exist.

Author: ```suggestion
Here we are going to make a directory or folder called `data`, if it doesn't already exist. No need for you to change this code. 
```

```

## Get the data

> We are going to use two different country datasets from Kaggle.

Author: ```suggestion
We are going to use two different country datasets from [Kaggle](https://www.kaggle.com/).
```

This UN data-derived dataset with various stats about different countries: https://www.kaggle.com/datasets/sudalairajkumar/undata-country-profiles
The resulting CSV file is called `country_profile_variables.csv`.

This literacy dataset with statistics about literacy across countries
https://www.kaggle.com/datasets/programmerrdai/literacy

The resulting CSV file is called `cross-country-literacy-rates`.

> Both datasets have been downloaded and placed in the `data` and `raw` folder. Use readr::read_csv function to reach each dataset in. Call the first one `un_df` and the second `literacy_df`.

Author: ```suggestion
Both datasets have been downloaded and placed in the `raw` folder (also called a directory) within the `data` folder of the `03_Cleaning_the_Data` project folder. Use `readr::read_csv()` function to read each dataset into R. Call the first one `un_df` and the second `literacy_df`. 
```

Both datasets have been downloaded and placed in the `raw` folder (also called a directory) within the `data` folder of the `03_Cleaning_the_Data` project folder. Use `readr::read_csv()` function to read each dataset into R. Call the first one `un_df` and the second `literacy_df`. 

```{r, solution = TRUE}
# Read in the dataset for un_df
### Your code here
```

```{r, solution = TRUE}
# Read in the dataset for literacy_df
### Your code here
```

Use functions we have discussed previously to see what variables are in both `un_df` and `us_gov_df`.

```{r, solution = TRUE}
### Your code here
```

```{r, solution = TRUE}
### Your code here
```

### Cleaning the datasets

> You'll notice both datasets have pretty messy looking column names that are annoying to use. Use the `janitor::clean_names()` function to clean these up.

Author: ```suggestion
You'll notice both datasets have pretty messy looking column names that are annoying to use. Use the `janitor::clean_names()` function to clean these up. Don't forget to reassign the data objects!
```

```{r, solution = TRUE}
### Your code here
```

#### Clean up literacy_df

The `literacy_df` data frame shows multiple rows for each country. For this analysis, let's only keep the latest year recorded for each country.
You can use  `arrange()` and `desc()` (or a minus sign)  to first order the data by the latest year recorded for each country.  Then use the `distinct()` function to keep only rows from the lastest year for each country by essentially selecting for the first mention of each country. Hint you will need to use an argument within `distinct()` to not lose the other columns. (Again don't forget to reassign your data objects to keep your work.)

```{r, solution = TRUE}
### Your code here

> # or

Author: ```suggestion
```

> ### Your code here

Author: ```suggestion
```

```

#### Clean up un_df

> There's a number of variables in `un_df` that are numeric but are being treated as a character.

Author: ```suggestion
There's a number of variables in `un_df` that are numeric but are being treated as a character. Use the `glimpse()` function of the `dplyr` package to take a look at the variables to determine how often there are numeric values that are being treated as a character.
```

Let's take a look at `life_expectancy_at_birth_females_males_years` as an example. Use the `head()` function and print out the first 10 rows to see.  

```{r, solution = TRUE}
### Your code here
```

Let's also take a closer look at `co2_emission_estimates_million_tons_tons_per_capita` as an example.

```{r, solution = TRUE}
### Your code here
```

> Two things we can notice here. One is that really it looks like this is two variables, one for female and one for male. We'll want to split this into two columns.

Author: ```suggestion
Two things we can notice here. One is that really `life_expectancy_at_birth_females_males_years` looks like it is two variables, one for female and one for male. We'll want to split this into two columns. 
```

> Another is that they are using two different items to note missing data `-99` and `...` For R to deal with missing data appopriately we will want to change these to be `NA`.

Author: ```suggestion
Another is that they are using two different items to note missing data `-99` and `...` For R to deal with missing data appropriately we will want to change these to be `NA`. This is also true for the `co2_emission_estimates_million_tons_tons_per_capita` variable which is being considered by R as a numeric value (based on how it is labeled as `dbl`).
```

> Let's start with the missing data. To replace these `-99` and `...` we can use the function: `naniar::replace_with_na()`. We'll want to do this for every variable, not just the one we looked at above.

Author: ```suggestion
Let's start with some of the missing data. To replace these `-99`(the numeric version) and `"-99"` (the character version) and `.../...` , we can use the function: `naniar::replace_with_na()`. We'll want to do this for every variable, not just the one we looked at above.  When we replace each of these they will have a single `NA` value even though we have two values for the observations where we have numeric values.
```

Let's run `summary()` on it.  

```{r, solution = TRUE}
### Your code here
```

The regions listed are repeated but because they are characters, summary doesn't give us useful information about the categories.
This `region` columns would be best treated as factors. Turn this `region` column into a factor.

```{r, solution = TRUE}
### Your code here
```

Re-run `summary` on your region columns  to confirm that the data make more sense now.

```{r, solution = TRUE}
### Your code here
```

### Joining the data

Now we have two generally clean datasets that both have information about countries. Use a `dplyr::join` function to join all the rows for countries that are in both `un_df` and `literacy_df` but exclude rows from countries that aren't in both. Look up `?dplyr::join` for more help. You will need to use the `by` argument.

Call this new data frame `countries_df`.

```{r, solution = TRUE}
### Your code here
```

We will be returning to this dataset in a future project. Save `countries_df` to an RDS file here. Save this file to the `data` folder and name it `countries_df.rds`.

```{r, solution = TRUE}
### Your code here
```

### Reshape data in preparation for plotting

> We want to see how literacy is related to life expectancy. But have literacy split up by gender. Let's reshape our data into something that will be easier to plot.

Author: ```suggestion
First, select only the following columns: `country`, `year`, `life_expectancy_at_birth_females`, `life_expectancy_at_birth_males`, and `literacy_rates_world_bank_cia_world_factbook_and_other_sources`. Name this new `select`ed data frame `plotting_df`. 
```
